explorer/ast/statement.h

```diff
@@ -298,6 +298,108 @@ class While : public Statement {
   Nonnull<Block*> body_;
 };
 
+class For : public Statement {
+ public:
+  For(SourceLocation source_loc, Nonnull<BindingPattern*> variable_declaration,
+      Nonnull<Expression*> loop_target, Nonnull<Block*> body)
+      : Statement(AstNodeKind::For, source_loc),
+        variable_declaration_(variable_declaration),
+        loop_target_(loop_target),
+        body_(body) {
+    Nonnull<Arena*> arena = new Arena();
+    auto index_name =
+        arena->New<IdentifierExpression>(source_loc, "$yy_forloop");
```

Comment:
We generally try to have the parsed AST match the syntax that the user supplied, because otherwise things like error messages and trace output can get very confusing, because they're talking about code that the user can't see. We've sometimes talked about implementing some features in terms of AST rewriting, but haven't figured out a design that we're happy with.

Instead of introducing these artificial AST nodes so that they can be fed to the interpreter, my recommendation would be to have logic in the part of the interpreter that implements `For` that directly does what execution of those nodes would do. For example, instead of synthesizing an AST node that calls `__intrinsic_array_sz`, have code in the interpreter that directly calls `.elements().vec().size()` on the array (as a side benefit, I think that means we don't need `__intrinsiz_array_sz` in the first place). Does that make sense?

Reply:
Ok! I will remove the code and try to implement as logic in the node.

Reply:
We may need or want to create additional AST nodes to represent the semantics of the  loop -- for example, if we have an interface for converting an arbitrary sequence into something we can iterate over. If you find that's necessary, that would be better done in type_checking.cpp when we type-check the statement, rather than here.